src/solvers/slvr_blk_2m.hpp

```diff
@@ -0,0 +1,399 @@
+#pragma once
+#include "slvr_sgs.hpp"
+
+#include <libcloudph++/blk_2m/options.hpp>
+#include <libcloudph++/blk_2m/rhs_cellwise.hpp>
+#include <libcloudph++/blk_2m/rhs_columnwise.hpp>
+
+
+template <class ct_params_t>
+class slvr_blk_2m_common : public std::conditional_t<ct_params_t::sgs_scheme == libmpdataxx::solvers::iles,
+                                                     slvr_common<ct_params_t>,
+                                                     slvr_sgs<ct_params_t>
+                                                    >
+{
+  using parent_t = std::conditional_t<ct_params_t::sgs_scheme == libmpdataxx::solvers::iles,
+                                      slvr_common<ct_params_t>,
+                                      slvr_sgs<ct_params_t>
+                                     >;
+
+  public:
+  using ix = typename ct_params_t::ix; // TODO: it's now in solver_common - is it needed here?
+  using real_t = typename ct_params_t::real_t;
+  using clock = typename parent_t::clock;
+  private:
+
+  // a 2D/3D array with copy of the environmental total pressure of dry air
+  typename parent_t::arr_t &p_e;
+
+  protected:
+
+  // accumulated water falling out of domain
+  real_t puddle;
+
+  void diag()
+  {
+    assert(this->rank == 0);
+    parent_t::tbeg = parent_t::clock::now();
+
+    // recording puddle
+    for(int i=0; i < 10; ++i)
+    {
+       this->f_puddle << i << " " << (i == 8 ? this->puddle : 0) << "\n";
+    }
+    this->f_puddle << "\n";
+
+    parent_t::tend = parent_t::clock::now();
+    parent_t::tdiag += std::chrono::duration_cast<std::chrono::milliseconds>( parent_t::tend - parent_t::tbeg );
+  }
+
+  void rc_src();
+  void nc_src();
+  void rr_src();
+  void nr_src();
+  bool get_rain() { return opts.acnv; }
+  void set_rain(bool val) {
+    opts.acnv = val;
+    opts.RH_max = val ? 44 : 1.01;
+  };
+
+  virtual typename parent_t::arr_t get_rc(typename parent_t::arr_t&) final
+    {
+  return this->state(ix::rc);
+    }
+
+  void record_all()
+  {
+    // plain (no xdmf) hdf5 output
+    parent_t::output_t::parent_t::record_all();
+    // UWLCM output
+    diag();
+    // xmf markup
+    this->write_xmfs();
+  }
+
+  void hook_mixed_rhs_ante_loop()
+  {}
+  void hook_mixed_rhs_ante_step()
+  {
+    const auto nc     = this->state(ix::nc)(this->ijk);
+    update_rhs(this->rhs, this->dt, 0);
+    negcheck(nc, "nc before apply rhs ante step");
+    this->apply_rhs(this->dt);
+  }
+  void hook_mixed_rhs_post_step()
+  {
+    const auto nc     = this->state(ix::nc)(this->ijk);
+    update_rhs(this->rhs, this->dt, 1);
+    this->apply_rhs(this->dt);
+  }
+
+  void hook_ante_loop(int nt)
+  {
+    // fill with zeros
+    this->state(ix::rc)(this->ijk) = 0;
+    this->state(ix::rr)(this->ijk) = 0;
+    this->state(ix::nc)(this->ijk) = 0;
+    this->state(ix::nr)(this->ijk) = 0;
```

Comment:
Shouldn't initial nc and nr be set by caseptr->intcond call?

Reply:
I'm not sure. I was replicating what is in slvr_blk_1m. https://github.com/igfuw/UWLCM/blob/61fb8962cf6f81e3d997b16f3e655a39e4ba2721/src/solvers/slvr_blk_1m.hpp#L105-L109

Reply:
Would prolly be better to set it via the initial condition. For bulk-1m scheme we can be a little more "lazy" because saturation adjustemnt will take care of rc at first time step. For 2-moment scheme it could cause some problems for cases that already start with liquid water present, such as dycoms.

Reply:
On second thought, bulk-2m is similar to lagrangian: if there is initial supersaturation, we don't know what should be initial nc (right?), so we set it to 0 and run spinup.